radsat-sk/operation/subsystems/RPdb.c

```diff
@@ -0,0 +1,336 @@
+
+/**
+ * 2 Byte Command for Resetting PDB Watchdog (0x2200)
+ */
+static uint8_t pdbWatchdogResetCommand[2] = {0x22, 0x00};
+
+
+/***************************************************************************************************
+                                       PRIVATE FUNCTION STUBS
+***************************************************************************************************/
+
+static int pdbTalk(uint8_t* command, uint8_t* response);
+
+
+/***************************************************************************************************
+                                             PUBLIC API
+***************************************************************************************************/
+/**
+ * Get the Irradiance (W/m^2) value from the sun sensor on each face of the CubeSat
+ *
+ * @param A pointer to a SunSensorStatus object, which will be filled with the converted ADC data for each sun sensor
+ * @return an error code, or 0 if no error
+ */
+int getSunSensorData(SunSensorStatus* sunData) {
+
+	// Create temporary variables for sending and receiving I2C data
+	uint8_t i2c_command[PDB_TELEM_COMMAND_LENGTH] = {0};
+	uint8_t i2c_data[PDB_RESPONSE_LENGTH] = {0};
+
+	// Create a temporary array for calculated sun data before transferring into the structure sunData
+	float convertedData[NUM_SUN_SENSORS] = {0};
+
+	// Send 6 commands to get each of the sun sensor's data
+	for(int i = 0; i < NUM_SUN_SENSORS; i = i + 1){
+		// Using address of the command into the i2c_command
+		// TODO: Remove this before push ^
+		// Load command into output buffer
+		memcpy(i2c_command, &pdbSunSensorCommandBytes[i], PDB_TELEM_COMMAND_LENGTH);
+		int error = pdbTalk(i2c_command, i2c_data);
+
+		if(error != 0){
+			// TODO: Some error raising here?
+			// return the error
+			return error;
+		}
+
+		memcpy(&convertedData[i], i2c_data, PDB_RESPONSE_LENGTH);
+		convertedData[i] = ADC_COUNT_TO_IRRADIANCE * ((float)convertedData[i]);
+	}
+
+	// Now store all of the calculated data into the proper slot in the sunData structure
+	sunData->xPos = convertedData[3]; 	// SA2B
+	sunData->xNeg = convertedData[2]; 	// SA2A
+	sunData->yPos = convertedData[0]; 	// SA1A
+	sunData->yNeg = convertedData[1]; 	// SA1B
+	sunData->zPos = convertedData[5]; 	// SA3B
+	sunData->zNeg = convertedData[4]; 	// SA3A
+	// Could change these numbers into an ENUM with the names of the connections
+
+	return 0;
+}
+
+
+/**
+ * Request and store all of the relevant data (voltage, current, temperature, sun sensor data)
+ *  	from the PDB board
+ *
+ * @param A pointer to an PDBStatus object containing all of the telemetry for the PDB system
+ * @return either 0 or an error code
+ */
+int getPdbTelemetry(PdbStatus* dataStorage) {
+
+	// Create an empty sunSensorData structure
+	SunSensorStatus sunSensorData;
+
+	// Get sun sensor data and store it in the new PDBStatus Structure
+	getSunSensorData(&sunSensorData);
+
+	dataStorage->sunSensorData = sunSensorData;
+
```

Comment:
If you're just going to assign it to `dataStorage->sunSensorData` immediately anyway, why not just pass that as a pointer directly? E.g. don't even use a temporary `sunSensorData` variable, and make the get call like so: `getSunSensorData(&dataStorage->sunSensorData);`

Reply:
I like this a lot better, didn't even think about it (or know it was possible haha). This method is way cleaner too.